pkg/cmd/provisioning/roles.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	// Ensure role name is no longer than 64 charactters
+	var shortenedRoleName string
+	if len(roleName) > 64 {
+		shortenedRoleName = roleName[0:64]
+	} else {
+		shortenedRoleName = roleName
+	}
+
+	rolePolicyDocument, err := createRolePolicyDocument(oidcProviderARN, issuerURL, credReq.Spec.ServiceAccountNames)
+	if err != nil {
+		return "", errors.Wrap(err, "Error while create Role policy document")
+	}
+
+	roleDescription := fmt.Sprintf("OpenShift role for %s/%s", credReq.Spec.SecretRef.Namespace, credReq.Spec.SecretRef.Name)
+
+	rolePolicy := createRolePolicy(awsProviderSpec.StatementEntries)
+
+	if generateOnly {
+		// Generate Role
+		// Need to escape all the double quotes in the generated JSON rolePolicyDocument
+		// so that the JSON text file is valid
+		escapedRolePolicyDoc := strings.Replace(rolePolicyDocument, "\"", "\\\"", -1)
+		roleJSON := fmt.Sprintf(roleTemplate, shortenedRoleName, escapedRolePolicyDoc, roleDescription)
+		roleFilename := fmt.Sprintf(roleFilenameFormat, roleNum, roleName)
+		roleFullPath := filepath.Join(targetDir, roleFilename)
+		if err := saveToFile(roleDescription, roleFullPath, []byte(roleJSON)); err != nil {
+			return "", errors.Wrap(err, "failed to save Role content JSON")
+		}
+
+		// Generate Role Policy
+		escapedRolePolicy := strings.Replace(rolePolicy, "\"", "\\\"", -1)
+		rolePolicyJSON := fmt.Sprintf(rolePolicyTemplate, shortenedRoleName, shortenedRoleName, escapedRolePolicy)
+		rolePolicyFilename := fmt.Sprintf(rolePolicyFilenameFormat, roleNum, roleName)
+		rolePolicyFullPath := filepath.Join(targetDir, rolePolicyFilename)
+		if err := saveToFile(roleDescription, rolePolicyFullPath, []byte(rolePolicyJSON)); err != nil {
+			return "", errors.Wrap(err, "failed to save Role Policy content JSON")
+		}
+
+	} else {
+		var role *iam.Role
+		outRole, err := awsClient.GetRole(&iam.GetRoleInput{
+			RoleName: awssdk.String(shortenedRoleName),
+		})
+
+		if err != nil {
+			var aerr awserr.Error
+			if errors.As(err, &aerr) {
+				switch aerr.Code() {
+				case iam.ErrCodeNoSuchEntityException:
+
+					roleOutput, err := awsClient.CreateRole(&iam.CreateRoleInput{
+						RoleName:                 awssdk.String(shortenedRoleName),
+						Description:              awssdk.String(roleDescription),
+						AssumeRolePolicyDocument: awssdk.String(rolePolicyDocument),
+						Tags: []*iam.Tag{
+							{
+								Key:   awssdk.String(fmt.Sprintf("%s/%s", ccoctlAWSResourceTagKeyPrefix, namePrefix)),
+								Value: awssdk.String(ownedCcoctlAWSResourceTagValue),
+							},
+						},
+					})
+					if err != nil {
+						return "", errors.Wrap(err, "Failed to create role")
+					}
+
+					role = roleOutput.Role
+					log.Printf("Role %s created", *role.Arn)
+
+				default:
+					return "", err
+				}
+
+			}
+		} else {
+			role = outRole.Role
+			log.Printf("Existing role %s found", *role.Arn)
+
+			// TODO: implement idemponent apply/update for role
```

Comment:
Future PR?

Reply:
Possibly. Doing a GetRolePolicy() and comparing what we get back to what we generate is a bit involved as AWS will reformat things so you can't do a straight string comparison. I just wanted to note here that if we were really motivated, we could avoid making this mutating call.